File: typescript/src/routes/users.test.ts
import { describe, it, expect, beforeEach } from 'vitest'
import app from '../index'
import { store } from '../stores'
import type { operations, components } from '../types/api'
import { loginTestUser, createAuthHeaders, TEST_USERS } from '../test-helpers/auth'

type User = components['schemas']['User']
type Cart = components['schemas']['Cart']
type ErrorResponse = components['schemas']['ErrorResponse']
type UserListResponse = {
  items: User[];
  total: number;
  limit: number;
  offset: number;
}

describe('Users API', () => {
  let authToken: string

  beforeEach(async () => {
    // Reset store to initial state before each test
    const storeInstance = new (store.constructor as any)()
    Object.setPrototypeOf(store, Object.getPrototypeOf(storeInstance))
    Object.assign(store, storeInstance)

    // Login to get auth token
    authToken = await loginTestUser(TEST_USERS.alice.email, TEST_USERS.alice.password)
  })

  describe('GET /users', () => {
    it('should return all users with authentication', async () => {
      const res = await app.request('/users', {
        headers: createAuthHeaders(authToken),
      })
      const json = await res.json() as UserListResponse

      expect(res.status).toBe(200)
      expect(json).toHaveProperty('items')
      expect(json).toHaveProperty('total')
      expect(json).toHaveProperty('limit')
      expect(json).toHaveProperty('offset')
      expect(Array.isArray(json.items)).toBe(true)
      expect(json.items).toHaveLength(2)
      expect(json.total).toBe(2)
      expect(json.limit).toBe(20)
      expect(json.offset).toBe(0)
      expect(json.items[0]).toHaveProperty('id')
      expect(json.items[0]).toHaveProperty('email')
      expect(json.items[0]).toHaveProperty('name')
      expect(json.items[0]).toHaveProperty('address')
    })

    it('should return 401 without authentication', async () => {
      const res = await app.request('/users')
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(401)
      expect(json.error).toHaveProperty('code', 'UNAUTHORIZED')
      expect(json.error).toHaveProperty('message')
    })

    it('should support pagination with authentication', async () => {
      const res1 = await app.request('/users?limit=1&offset=0', {
        headers: createAuthHeaders(authToken),
      })
      const json1 = await res1.json() as UserListResponse

      expect(res1.status).toBe(200)
      expect(json1.items).toHaveLength(1)
      expect(json1.total).toBe(2)
      expect(json1.limit).toBe(1)
      expect(json1.offset).toBe(0)

      const res2 = await app.request('/users?limit=1&offset=1', {
        headers: createAuthHeaders(authToken),
      })
      const json2 = await res2.json() as UserListResponse

      expect(res2.status).toBe(200)
      expect(json2.items).toHaveLength(1)
      expect(json2.total).toBe(2)
      expect(json2.limit).toBe(1)
      expect(json2.offset).toBe(1)

      // Ensure different users
      expect(json1.items[0].id).not.toBe(json2.items[0].id)
    })
  })

  describe('GET /users/:id', () => {
    it('should return a user by id with authentication', async () => {
      const res = await app.request('/users/1', {
        headers: createAuthHeaders(authToken),
      })
      const json = await res.json() as User | ErrorResponse

      expect(res.status).toBe(200)
      if ('id' in json) {
        expect(json.id).toBe('1')
        expect(json.email).toBe('user1@example.com')
        expect(json.name).toBe('Test User 1')
        expect(json.address).toEqual({
          street: '123 Test St',
          city: 'Test City',
          state: 'TC',
          postalCode: '12345',
          country: 'USA',
        })
      }
    })

    it('should return 401 without authentication', async () => {
      const res = await app.request('/users/1')
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(401)
      expect(json.error).toHaveProperty('code', 'UNAUTHORIZED')
    })

    it('should return 404 for non-existent user with authentication', async () => {
      const res = await app.request('/users/999', {
        headers: createAuthHeaders(authToken),
      })
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(404)
      expect(json.error).toHaveProperty('code', 'NOT_FOUND')
      expect(json.error).toHaveProperty('message')
    })
  })

  describe('POST /users', () => {
    it('should create a new user with address and authentication', async () => {
      const newUser = {
        email: 'newuser@example.com',
        name: 'New User',
        address: {
          street: '789 New St',
          city: 'New City',
          state: 'NC',
          postalCode: '11111',
          country: 'USA'
        }
      }

      const res = await app.request('/users', {
        method: 'POST',
        headers: {
          ...createAuthHeaders(authToken),
        },
        body: JSON.stringify(newUser),
      })
      const json = await res.json() as User

      expect(res.status).toBe(201)
      expect(json).toHaveProperty('id')
      expect(json.email).toBe(newUser.email)
      expect(json.name).toBe(newUser.name)
      expect(json.address).toEqual(newUser.address)
      expect(json).toHaveProperty('createdAt')
      expect(json).toHaveProperty('updatedAt')

      // Verify cart was created for new user
      const cartRes = await app.request(`/carts/users/${json.id}`, {
        headers: createAuthHeaders(authToken),
      })
      const cart = await cartRes.json() as Cart
      expect(cart).toHaveProperty('userId', json.id)
      expect(cart.items).toHaveLength(0)
    })

    it('should create a new user without address with authentication', async () => {
      const newUser = {
        email: 'minimal@example.com',
        name: 'Minimal User'
      }

      const res = await app.request('/users', {
        method: 'POST',
        headers: {
          ...createAuthHeaders(authToken),
        },
        body: JSON.stringify(newUser),
      })
      const json = await res.json() as User

      expect(res.status).toBe(201)
      expect(json).toHaveProperty('id')
      expect(json.email).toBe(newUser.email)
      expect(json.name).toBe(newUser.name)
      expect(json.address).toBeUndefined()
    })

    it('should return 401 without authentication', async () => {
      const newUser = {
        email: 'unauth@example.com',
        name: 'Unauthorized User'
      }

      const res = await app.request('/users', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(newUser),
      })
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(401)
      expect(json.error).toHaveProperty('code', 'UNAUTHORIZED')
    })
  })

  describe('PUT /users/:id', () => {
    it('should update user details with authentication', async () => {
      const updateData = {
        name: 'Updated User Name',
        email: 'updated@example.com'
      }

      const res = await app.request('/users/1', {
        method: 'PUT',
        headers: {
          ...createAuthHeaders(authToken),
        },
        body: JSON.stringify(updateData),
      })
      const json = await res.json() as User | ErrorResponse

      expect(res.status).toBe(200)
      if ('id' in json) {
        expect(json.id).toBe('1')
        expect(json.name).toBe(updateData.name)
        expect(json.email).toBe(updateData.email)
        expect(json).toHaveProperty('updatedAt')
      }
    })

    it('should update user address with authentication', async () => {
      const updateData = {
        address: {
          street: '999 Updated Ave',
          city: 'Update City',
          state: 'UC',
          postalCode: '99999',
          country: 'USA'
        }
      }

      const res = await app.request('/users/1', {
        method: 'PUT',
        headers: {
          ...createAuthHeaders(authToken),
        },
        body: JSON.stringify(updateData),
      })
      const json = await res.json() as User | ErrorResponse

      expect(res.status).toBe(200)
      if ('id' in json) {
        expect(json.address).toEqual(updateData.address)
      }
    })

    it('should return 401 without authentication', async () => {
      const updateData = {
        name: 'Unauthorized Update'
      }

      const res = await app.request('/users/1', {
        method: 'PUT',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(updateData),
      })
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(401)
      expect(json.error).toHaveProperty('code', 'UNAUTHORIZED')
    })

    it('should return 404 when updating non-existent user with authentication', async () => {
      const updateData = {
        name: 'Non-existent'
      }

      const res = await app.request('/users/999', {
        method: 'PUT',
        headers: {
          ...createAuthHeaders(authToken),
        },
        body: JSON.stringify(updateData),
      })
      const json = await res.json() as ErrorResponse

      expect(res.status).toBe(404)
      expect(json.error).toHaveProperty('code', 'NOT_FOUND')
    })
  })

  describe('DELETE /users/:id', () => {
    it('should delete a user with authentication', async () => {
      const res = await app.request('/users/2', {
        method: 'DELETE',
        headers: createAuthHeaders(authToken),
      })

      expect(res.status).toBe(204)
      expect(await res.text()).toBe('')

      // Verify deletion
      const getRes = await app.request('/users/2', {
        headers: createAuthHeaders(authToken),
      })
      expect(getRes.status).toBe(404)
    })

    it('should return 401 without authentication', async () => {
      const res = await app.request('/users/2', {
        method: 'DELETE',
      })

      expect(res.status).toBe(401)
      const json = await res.json() as ErrorResponse
      expect(json.error).toHaveProperty('code', 'UNAUTHORIZED')
    })

    it('should return 404 when deleting non-existent user with authentication', async () => {
      const res = await app.request('/users/999', {
        method: 'DELETE',
        headers: createAuthHeaders(authToken),
      })

      expect(res.status).toBe(404)
    })
  })
})

File: go/internal/handlers/users_test.go
package handlers_test

import (
	"net/http"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestUsersService_List(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should return all users with default pagination", func(t *testing.T) {
		rr := makeAuthenticatedRequest(t, server, "GET", "/users", nil, token)

		assertStatus(t, rr, http.StatusOK)
		response := assertPaginatedResponse(t, rr, 2, 20, 0) // 2 default users

		items := response["items"].([]any)
		assert.Len(t, items, 2)

		// Check first user structure
		user := items[0].(map[string]any)
		assert.Contains(t, user, "id")
		assert.Contains(t, user, "email")
		assert.Contains(t, user, "name")
		assert.Contains(t, user, "address")
		assert.Contains(t, user, "createdAt")
		assert.Contains(t, user, "updatedAt")
	})

	t.Run("should support pagination with limit and offset", func(t *testing.T) {
		// First page
		rr1 := makeAuthenticatedRequest(t, server, "GET", "/users?limit=1&offset=0", nil, token)
		assertStatus(t, rr1, http.StatusOK)
		response1 := assertPaginatedResponse(t, rr1, 2, 1, 0)
		items1 := response1["items"].([]any)
		assert.Len(t, items1, 1)

		// Second page
		rr2 := makeAuthenticatedRequest(t, server, "GET", "/users?limit=1&offset=1", nil, token)
		assertStatus(t, rr2, http.StatusOK)
		response2 := assertPaginatedResponse(t, rr2, 2, 1, 1)
		items2 := response2["items"].([]any)
		assert.Len(t, items2, 1)

		// Ensure different users
		user1 := items1[0].(map[string]any)
		user2 := items2[0].(map[string]any)
		assert.NotEqual(t, user1["id"], user2["id"])
	})

	t.Run("should handle empty results with offset beyond total", func(t *testing.T) {
		rr := makeAuthenticatedRequest(t, server, "GET", "/users?limit=10&offset=100", nil, token)
		assertStatus(t, rr, http.StatusOK)
		response := assertPaginatedResponse(t, rr, 2, 10, 100)
		items := response["items"].([]any)
		assert.Len(t, items, 0)
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		rr := makeRequest(t, server, "GET", "/users", nil)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestUsersService_Get(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should return a user by id", func(t *testing.T) {
		rr := makeAuthenticatedRequest(t, server, "GET", "/users/1", nil, token)
		assertStatus(t, rr, http.StatusOK)

		var user map[string]any
		err := decodeJSON(rr, &user)
		require.NoError(t, err)

		assert.Equal(t, "1", user["id"])
		assert.Equal(t, "user1@example.com", user["email"])
		assert.Equal(t, "Test User 1", user["name"])
		assert.NotNil(t, user["address"])
	})

	t.Run("should return 404 for non-existent user", func(t *testing.T) {
		rr := makeAuthenticatedRequest(t, server, "GET", "/users/999", nil, token)
		assertStatus(t, rr, http.StatusNotFound)
		assertErrorResponse(t, rr, "NOT_FOUND")
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		rr := makeRequest(t, server, "GET", "/users/1", nil)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestUsersService_Create(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should create a new user with address", func(t *testing.T) {
		newUser := map[string]any{
			"email": "newuser@example.com",
			"name":  "New User",
			"address": map[string]any{
				"street":     "789 New St",
				"city":       "New City",
				"state":      "NC",
				"postalCode": "67890",
				"country":    "USA",
			},
		}

		rr := makeAuthenticatedRequest(t, server, "POST", "/users", newUser, token)
		assertStatus(t, rr, http.StatusCreated)

		var user map[string]any
		err := decodeJSON(rr, &user)
		require.NoError(t, err)

		assert.NotEmpty(t, user["id"])
		assert.Equal(t, newUser["email"], user["email"])
		assert.Equal(t, newUser["name"], user["name"])
		assert.NotNil(t, user["address"])
		assert.NotEmpty(t, user["createdAt"])
		assert.NotEmpty(t, user["updatedAt"])

		// Verify cart was created for new user
		userID := user["id"].(string)
		cartRR := makeAuthenticatedRequest(t, server, "GET", "/carts/users/"+userID, nil, token)
		assertStatus(t, cartRR, http.StatusOK)

		var cart map[string]any
		err = decodeJSON(cartRR, &cart)
		require.NoError(t, err)
		assert.Equal(t, userID, cart["userId"])
		assert.Empty(t, cart["items"])
	})

	t.Run("should create a new user without address", func(t *testing.T) {
		newUser := map[string]any{
			"email": "minimal@example.com",
			"name":  "Minimal User",
		}

		rr := makeAuthenticatedRequest(t, server, "POST", "/users", newUser, token)
		assertStatus(t, rr, http.StatusCreated)

		var user map[string]any
		err := decodeJSON(rr, &user)
		require.NoError(t, err)

		assert.NotEmpty(t, user["id"])
		assert.Equal(t, newUser["email"], user["email"])
		assert.Equal(t, newUser["name"], user["name"])
		assert.Nil(t, user["address"])
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		newUser := map[string]any{
			"email": "unauth@example.com",
			"name":  "Unauthorized User",
		}

		rr := makeRequest(t, server, "POST", "/users", newUser)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})

	// TODO: Implement validation in Go handlers
	// t.Run("should return 400 for invalid user data", func(t *testing.T) {
	// 	invalidUser := map[string]any{
	// 		"email": "", // Empty email
	// 		"name":  "Invalid User",
	// 	}

	// 	rr := makeAuthenticatedRequest(t, server, "POST", "/users", invalidUser, token)
	// 	assertStatus(t, rr, http.StatusBadRequest)
	// 	assertErrorResponse(t, rr, "VALIDATION_ERROR")
	// })
}

func TestUsersService_Update(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should update user details", func(t *testing.T) {
		update := map[string]any{
			"name": "Updated Name",
		}

		rr := makeAuthenticatedRequest(t, server, "PATCH", "/users/1", update, token)
		assertStatus(t, rr, http.StatusOK)

		var user map[string]any
		err := decodeJSON(rr, &user)
		require.NoError(t, err)

		assert.Equal(t, "1", user["id"])
		assert.Equal(t, "Updated Name", user["name"])
		assert.Equal(t, "user1@example.com", user["email"]) // Email unchanged
	})

	t.Run("should update user address", func(t *testing.T) {
		update := map[string]any{
			"address": map[string]any{
				"street":     "999 Updated St",
				"city":       "Updated City",
				"state":      "UC",
				"postalCode": "99999",
				"country":    "USA",
			},
		}

		rr := makeAuthenticatedRequest(t, server, "PATCH", "/users/1", update, token)
		assertStatus(t, rr, http.StatusOK)

		var user map[string]any
		err := decodeJSON(rr, &user)
		require.NoError(t, err)

		address := user["address"].(map[string]any)
		assert.Equal(t, "999 Updated St", address["street"])
		assert.Equal(t, "Updated City", address["city"])
	})

	t.Run("should return 404 when updating non-existent user", func(t *testing.T) {
		update := map[string]any{
			"name": "Ghost User",
		}

		rr := makeAuthenticatedRequest(t, server, "PATCH", "/users/999", update, token)
		assertStatus(t, rr, http.StatusNotFound)
		assertErrorResponse(t, rr, "NOT_FOUND")
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		update := map[string]any{
			"name": "Unauthorized Update",
		}

		rr := makeRequest(t, server, "PATCH", "/users/1", update)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestUsersService_Delete(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should delete a user", func(t *testing.T) {
		// Create a new user to delete (note: createTestUser doesn't use auth)
		// We'll create using authenticated request instead
		newUser := map[string]any{
			"email": "delete@example.com",
			"name":  "Delete Me",
			"address": map[string]any{
				"street":     "123 Test St",
				"city":       "Test City",
				"state":      "TC",
				"postalCode": "12345",
				"country":    "Test Country",
			},
		}

		createRR := makeAuthenticatedRequest(t, server, "POST", "/users", newUser, token)
		assertStatus(t, createRR, http.StatusCreated)

		var user map[string]any
		err := decodeJSON(createRR, &user)
		require.NoError(t, err)
		userID := user["id"].(string)

		// Delete the user
		rr := makeAuthenticatedRequest(t, server, "DELETE", "/users/"+userID, nil, token)
		assertStatus(t, rr, http.StatusNoContent)

		// Verify user is deleted
		getRR := makeAuthenticatedRequest(t, server, "GET", "/users/"+userID, nil, token)
		assertStatus(t, getRR, http.StatusNotFound)
	})

	t.Run("should return 404 when deleting non-existent user", func(t *testing.T) {
		rr := makeAuthenticatedRequest(t, server, "DELETE", "/users/999", nil, token)
		assertStatus(t, rr, http.StatusNotFound)
		assertErrorResponse(t, rr, "NOT_FOUND")
	})

	t.Run("should return 401 without authentication", func(t *testing.T) {
		rr := makeRequest(t, server, "DELETE", "/users/1", nil)
		assertStatus(t, rr, http.StatusUnauthorized)
		assertErrorResponse(t, rr, "UNAUTHORIZED")
	})
}

func TestUsersService_Integration(t *testing.T) {
	server, _, token := setupTestServerWithAuth(t)

	t.Run("should handle full user lifecycle", func(t *testing.T) {
		// Create user
		newUser := map[string]any{
			"email": "lifecycle@example.com",
			"name":  "Lifecycle User",
			"address": map[string]any{
				"street":     "123 Test St",
				"city":       "Test City",
				"state":      "TC",
				"postalCode": "12345",
				"country":    "Test Country",
			},
		}

		createRR := makeAuthenticatedRequest(t, server, "POST", "/users", newUser, token)
		assertStatus(t, createRR, http.StatusCreated)

		var user map[string]any
		err := decodeJSON(createRR, &user)
		require.NoError(t, err)
		userID := user["id"].(string)

		// Update user
		update := map[string]any{
			"name": "Updated Lifecycle User",
		}
		updateRR := makeAuthenticatedRequest(t, server, "PATCH", "/users/"+userID, update, token)
		assertStatus(t, updateRR, http.StatusOK)

		// Verify update
		getRR := makeAuthenticatedRequest(t, server, "GET", "/users/"+userID, nil, token)
		assertStatus(t, getRR, http.StatusOK)

		var updatedUser map[string]any
		err = decodeJSON(getRR, &updatedUser)
		require.NoError(t, err)
		assert.Equal(t, "Updated Lifecycle User", updatedUser["name"])

		// Delete user
		deleteRR := makeAuthenticatedRequest(t, server, "DELETE", "/users/"+userID, nil, token)
		assertStatus(t, deleteRR, http.StatusNoContent)

		// Verify deletion
		getDeletedRR := makeAuthenticatedRequest(t, server, "GET", "/users/"+userID, nil, token)
		assertStatus(t, getDeletedRR, http.StatusNotFound)
	})
}
